File: src/exercicio-computacional-1.js
"use strict";

const fs = require("fs");

function sleep(ms) {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(ms);
    }, ms);
  });
}

function carregaArquivo(filename) {
  const conteudoLabirinto = fs.readFileSync(filename, "utf-8");
  const arrayLinhas = conteudoLabirinto.split("\n");

  const [tamanhoLinhas, tamanhoColunas] = arrayLinhas[0].split(" ").map(Number);

  const matrixLabirinto = [];
  let entradaLabirinto = null;

  // encontrar entrada
  for (let i = 1; i <= tamanhoLinhas; i++) {
    const linha = arrayLinhas[i].split("");
    matrixLabirinto.push(linha);

    if (!entradaLabirinto) {
      const posicaoEntradaLabirinto = linha.indexOf("e");
      if (posicaoEntradaLabirinto !== -1) {
        entradaLabirinto = [i - 1, posicaoEntradaLabirinto];
      }
    }
  }

  return { matrixLabirinto, tamanhoLinhas, tamanhoColunas, entradaLabirinto };
}

async function printMatriz(matrixLabirinto) {
  await sleep(100);
  console.clear();
  matrixLabirinto.forEach((linha) => console.log(linha.join("")));
}

async function pesquisaSaida({
  matrixLabirinto,
  tamanhoLinhas,
  tamanhoColunas,
  entradaLabirinto,
}) {
  const posicaoAtual = [entradaLabirinto];

  while (posicaoAtual.length > 0) {
    const [x, y] = posicaoAtual.pop();

    if (matrixLabirinto[x][y] === "s") {
      console.log("Saída encontrada!");
      console.log("Posição da saida:", x, y, "\n");

      matrixLabirinto[x][y] = "o";
      await printMatriz(matrixLabirinto);

      return true;
    }

    // Marca a posição atual como visitada
    if (matrixLabirinto[x][y] === "x" || matrixLabirinto[x][y] === "e") {
      matrixLabirinto[x][y] = "o";
    }

    await printMatriz(matrixLabirinto);

    const direcoes = [
      [x - 1, y], // cima
      [x + 1, y], // baixo
      [x, y - 1], // esquerda
      [x, y + 1], // direita
    ];

    for (const [nx, ny] of direcoes) {
      if (
        nx >= 0 &&
        ny >= 0 &&
        nx < tamanhoLinhas &&
        ny < tamanhoColunas &&
        (matrixLabirinto[nx][ny] === "x" || matrixLabirinto[nx][ny] === "s")
      ) {
        matrixLabirinto[x][y] = ".";
        posicaoAtual.push([nx, ny]);
      }
    }
  }

  console.log("Nenhum caminho para a saída foi encontrado.");

  return false;
}

async function main() {
  const matriz = carregaArquivo("../data/maze.txt");
  await printMatriz(matriz.matrixLabirinto);

  console.log("Resolvendo labirinto...\n");

  await pesquisaSaida(matriz);
}

(async () => {
  await main();
})();

File: src/exercicio-computacional-1.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <stack>
#include <thread>
#include <chrono>

using namespace std;

// Função para carregar o labirinto a partir de um arquivo
bool carregar_labirinto(const string& caminho_arquivo, vector<vector<char>>& labirinto, int& linhas, int& colunas, pair<int, int>& inicio) {
    ifstream arquivo(caminho_arquivo);
    if (!arquivo.is_open()) {
        cerr << "Erro ao abrir o arquivo." << endl;
        return false;
    }

    arquivo >> linhas >> colunas;
    labirinto.resize(linhas);
    string linha;
    getline(arquivo, linha); // Para ler o '\n' após as dimensões

    for (int i = 0; i < linhas; i++) {
        getline(arquivo, linha);
        labirinto[i] = vector<char>(linha.begin(), linha.end());
        size_t pos = linha.find('e');
        if (pos != string::npos) {
            inicio = {i, static_cast<int>(pos)};
        }
    }

    arquivo.close();
    return true;
}

// Função para imprimir o labirinto
void imprimir_labirinto(const vector<vector<char>>& labirinto) {
    this_thread::sleep_for(chrono::milliseconds(100));

    // Limpa a tela no terminal
    #ifdef _WIN32
        system("cls");
    #else
        system("clear");
    #endif

    for (const auto& linha : labirinto) {
        for (const auto& celula : linha) {
            cout << celula;
        }
        cout << endl;
    }
}

// Função para resolver o labirinto
bool resolver_labirinto(vector<vector<char>>& labirinto, int linhas, int colunas, pair<int, int> inicio) {
    stack<pair<int, int>> pilha;
    pilha.push(inicio);

    while (!pilha.empty()) {
        auto [x, y] = pilha.top();
        pilha.pop();

        // Verifica se é a saída
        if (labirinto[x][y] == 's') {
            labirinto[x][y] = 'o';
            cout << "Saída encontrada!\n";
            imprimir_labirinto(labirinto);
            return true;
        }

        if (labirinto[x][y] == 'x' || labirinto[x][y] == 'e') {
            labirinto[x][y] = 'o';
        }

        imprimir_labirinto(labirinto);
        cout << endl;

        labirinto[x][y] = '.';

        // Movimentos para as próximas posições (cima, baixo, esquerda, direita)
        for (const auto& [nx, ny] : vector<pair<int, int>>{{x - 1, y}, {x + 1, y}, {x, y - 1}, {x, y + 1}}) {
            if (nx >= 0 && nx < linhas && ny >= 0 && ny < colunas && labirinto[nx][ny] != '#' && labirinto[nx][ny] != '.') {
                pilha.push({nx, ny});
            }
        }
    }

    cout << "Nenhum caminho para a saída foi encontrado." << endl;
    return false;
}

int main() {
    vector<vector<char>> labirinto;
    int linhas, colunas;
    pair<int, int> inicio;

    if (!carregar_labirinto("../data/maze2.txt", labirinto, linhas, colunas, inicio)) {
        return 1;
    }

    imprimir_labirinto(labirinto);
    cout << "Resolvendo labirinto...\n";

    resolver_labirinto(labirinto, linhas, colunas, inicio);

    return 0;
}
